File: cpp/sha256.cpp
#include <openssl/sha.h>
#include "framework.hpp"

class Sha256: public HashBenchmark {
    private:
        static const int hashLen = 32;
        inline void _hashInternal(const std::string &password, unsigned char *hash) {
            SHA256(reinterpret_cast<const unsigned char*>(password.data()), password.size(), hash);
        }
    public:
        Sha256(std::string name) : HashBenchmark(name) {}

        std::string _hash(const std::string &password) {
            unsigned char hash[hashLen];
            _hashInternal(password, hash);
            std::string res;
            res.reserve(hashLen * 2);
            hexify(hash, hashLen, res);
            return res;
        }

        bool _checkHash(const std::string &hash, const std::string &password) {
            return hash == _hash(password);
        }
};

File: py/sha256.py
import hashlib

import framework

class Sha256(framework.HashBenchmark):
    def __init__(self, name, *args, **kwargs):
        super().__init__(name, *args, **kwargs)
    
    # Hash a password and return the string representation
    def _hash(self, password):
        h = hashlib.new('sha256')
        h.update(password.encode('utf-8'))
        return h.hexdigest()
    
    # Check if a hash matches a password
    def _checkHash(self, hash, password):
        return self._hash(password) == hash